lib/events.js

```diff
@@ -1,84 +1,129 @@
-      const hint = 'Hint: avoid adding listeners in loops';
-      console.warn(`${name}: ${warn}. ${max}. ${hint}`);
-    }
+  on(eventName, listener) {
+    const { events, maxListenersCount } = this;
+    if (!events.has(eventName)) events.set(eventName, new Set());
+    const listeners = this.events.get(eventName);
+    listeners.add(listener);
+    if (listeners.size <= maxListenersCount) return;
+    const title = 'MaxListenersExceededWarning';
+    const warn = 'Possible EventEmitter memory leak detected';
+    const max = `Current maxListenersCount is ${maxListenersCount}`;
+    const hint = 'Hint: avoid adding listeners in loops';
+    console.warn(`${title}: ${warn}. ${max}. ${hint}`);
   }
 
-  once(name, fn) {
-    const dispose = (...args) => {
-      this.off(name, dispose);
-      return void fn(...args);
+  once(eventName, listener) {
+    const wrapper = (...args) => {
+      this.off(eventName, wrapper);
+      return listener(...args);
     };
-    this.on(name, dispose);
+    wrapper.origin = listener;
+    this.on(eventName, wrapper);
+    this.wrappers.set(listener, wrapper);
   }
 
-  off(name, fn) {
-    const listeners = this.events.get(name);
-    if (!listeners) return;
-    const index = listeners.indexOf(fn);
-    listeners.splice(index, 1);
-    if (listeners.length === 0) {
-      this.events.delete(name);
-    }
+  emit(eventName, ...args) {
+    const listeners = this.events.get(eventName);
+    if (!listeners) return Promise.resolve();
+    const promises = [...listeners].map((listener) => listener(...args));
+    return Promise.all(promises).then(() => {});
   }
 
-  toPromise(name) {
-    return new Promise((resolve) => {
-      this.once(name, resolve);
-    });
+  off(eventName, listener) {
+    if (!listener) return void this.clear(eventName);
+    const listeners = this.events.get(eventName);
+    if (!listeners) return;
+    listeners.delete(listener);
+    const wrapped = this.wrappers.get(listener);
+    if (wrapped) {
+      listeners.delete(wrapped);
+      this.wrappers.delete(listener);
+    }
+    if (listeners.size === 0) this.events.delete(eventName);
   }
 
-  toIterator(name) {
-    const stub = () => {};
-    this.events.on(name, stub);
-    return {
-      async next() {},
-    };
+  clear(eventName) {
+    const { events, wrappers } = this;
+    if (!eventName) {
+      events.clear();
+      return void wrappers.clear();
+    }
+    const listeners = events.get(eventName);
+    if (!listeners) return;
+    for (const listener of listeners) {
+      wrappers.delete(listener);
+    }
+    events.delete(eventName);
   }
 
-  clear(name) {
-    if (!name) return void this.events.clear();
-    this.events.delete(name);
+  toPromise(eventName, options = {}) {
+    const { signal = null } = options;
+    return new Promise((resolve, reject) => {
+      if (!signal) {
+        return void this.once(eventName, (...args) => void resolve(args));
```

Comment:
Need use cases because we can't imagine all developer experience without textual representation